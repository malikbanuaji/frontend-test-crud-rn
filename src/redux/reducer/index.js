import ContactReducer from './contact';

export {ContactReducer};
