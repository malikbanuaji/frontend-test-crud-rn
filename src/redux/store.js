const {createStore, combineReducers} = require('redux');
const {ContactReducer} = require('./reducer');

function configureStore() {
  const rootReducer = combineReducers({
    contact: ContactReducer,
  });
  const store = createStore(rootReducer);
  return store;
}

export {configureStore};
