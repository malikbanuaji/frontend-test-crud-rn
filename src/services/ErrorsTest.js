class ResponseError extends Error {
  constructor(response = {}) {
    // Pass remaining arguments (including vendor specific ones) to parent constructor
    super(response);

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ResponseError);
    }

    this.name = 'ResponseError';
    // Custom debugging information
    this.message = response;
  }
}

module.exports = {ResponseError};
