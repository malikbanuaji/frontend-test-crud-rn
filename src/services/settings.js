export const TIMEOUT = 5000;
export const BASE_URL = 'https://simple-contact-crud.herokuapp.com';
