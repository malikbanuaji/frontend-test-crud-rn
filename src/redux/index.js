import {configureStore} from './store';

export {configureStore};
