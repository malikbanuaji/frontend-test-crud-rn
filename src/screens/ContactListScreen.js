import React, {useCallback, useEffect, useState} from 'react';
import {StatusBar, StyleSheet, View} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import ActionAddContact from '../components/common/ActionAddContact';
import HeaderBar from '../components/common/HeaderBar';
import HeaderBarTitle from '../components/common/HeaderBarTitle';
import ContactList from '../components/contact/ContactList';
import {setContactList} from '../redux/actions/contact';
import {Api} from '../services';
import {Colors, Spacing} from '../styles';
import Utils from '../utils';

export default function ContactListScreen({navigation}) {
  const [isLoading, setIsLoading] = useState(false);

  const _handlePressContactListItem = useCallback(
    (item) => {
      navigation.navigate('ContactDetail', {id: item.id});
    },
    [navigation],
  );
  const dispatch = useDispatch();

  const fetchAllContact = useCallback(async () => {
    setIsLoading(true);
    try {
      const response = await Api.fetchAllContact();
      dispatch(setContactList({data: response.data.data}));
      setIsLoading(false);
    } catch (error) {
      setIsLoading(false);
      Utils.SmallMessage.showError();
    }
  }, [dispatch]);

  useEffect(() => {
    fetchAllContact();
  }, [fetchAllContact]);

  const contactList = useSelector((state) => state.contact.data);

  function _handlePressAddContact() {
    navigation.navigate('CreateEditContact');
  }

  return (
    <View style={styles.container}>
      <StatusBar backgroundColor={Colors.primary} barStyle={'dark-content'} />
      <HeaderBar isLoading={false}>
        <HeaderBarTitle text={'KontakMu'} />
      </HeaderBar>
      <ContactList
        refreshing={isLoading}
        onRefresh={fetchAllContact}
        data={contactList}
        onPressItem={_handlePressContactListItem}
      />
      <View style={styles.actionContainer}>
        <ActionAddContact onPress={_handlePressAddContact} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
  },
  actionContainer: {
    position: 'absolute',
    right: Spacing.base,
    bottom: Spacing.base,
  },
});
