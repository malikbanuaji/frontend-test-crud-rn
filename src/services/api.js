import * as Settings from './settings';
import Axios from 'axios';

const config = {
  baseURL: Settings.BASE_URL,
  timeout: Settings.TIMEOUT,
};

export async function fetchAllContact() {
  return await Axios.get('/contact', config);
}

export async function fetchContact(id) {
  return await Axios.get(`/contact/${id}`, config);
}

export async function addContact({firstName, lastName, age, photo}) {
  return await Axios.post(
    '/contact',
    {firstName, lastName, age, photo},
    config,
  );
}

export async function editContact({id, firstName, lastName, age, photo}) {
  return await Axios.put(
    `/contact/${id}`,
    {firstName, lastName, age, photo},
    config,
  );
}

export async function deleteContact(id) {
  return await Axios.delete(`/contact/${id}`, config);
}
