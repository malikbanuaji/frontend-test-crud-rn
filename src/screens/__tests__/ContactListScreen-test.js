import {
  cleanup,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react-native';
import React from 'react';
import {Platform} from 'react-native';
import {Provider} from 'react-redux';
import {configureStore} from '../../redux';
import {Api} from '../../services';
import ContactListScreen from '../ContactListScreen';

jest.mock('../../services');

describe('ContactList', () => {
  let useEffect;
  let navigation;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f) => f());
  };

  const store = configureStore();

  beforeEach(() => {
    Platform.OS = 'android';

    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect();
    Api.fetchAllContact.mockResolvedValue({
      data: {
        data: [
          {
            id: '93ad6070-c92b-11e8-b02f-cbfa15db428b',
            firstName: 'Frodo',
            lastName: 'Baggins',
            age: 24,
            photo: 'N/A',
          },
        ],
      },
    });

    navigation = {
      navigate: jest.fn(),
    };
  });

  afterEach(cleanup);

  it('loads all contacts on start', async () => {
    const component = (
      <Provider store={store}>
        <ContactListScreen navigation={navigation} />
      </Provider>
    );

    const {getByTestId} = render(component);
    await waitFor(() => getByTestId('contact-list'));
    expect(Api.fetchAllContact).toBeCalled();
  });

  it('loads render all contacts on start', async () => {
    const component = (
      <Provider store={store}>
        <ContactListScreen navigation={navigation} />
      </Provider>
    );

    const {getByTestId} = render(component);
    const contactList = await waitFor(() => getByTestId('contact-list'));
    expect(contactList.props.data).toHaveLength(1);
  });

  it('handle button CreateEditContact correctly', async () => {
    const component = (
      <Provider store={store}>
        <ContactListScreen navigation={navigation} />
      </Provider>
    );

    const {getByTestId} = render(component);

    const addContactbutton = await waitFor(() =>
      getByTestId('action-add-contact-button'),
    );

    fireEvent.press(addContactbutton);
    expect(navigation.navigate).toBeCalledWith('CreateEditContact');
  });

  it('render correctly', async () => {
    const component = (
      <Provider store={store}>
        <ContactListScreen navigation={navigation} />
      </Provider>
    );

    const {getByTestId, toJSON} = render(component);
    await waitFor(() => getByTestId('contact-list'));
    expect(toJSON()).toMatchSnapshot();
  });

  it('handle press listItem correctly', async () => {
    const component = (
      <Provider store={store}>
        <ContactListScreen navigation={navigation} />
      </Provider>
    );

    const {getByTestId} = render(component);
    const contactList = await waitFor(() => getByTestId('contact-list'));

    const contactListItem = contactList.findByProps({
      testID: 'contact-list-item-button',
    });
    fireEvent.press(contactListItem);
    expect(navigation.navigate).toBeCalledWith(
      'ContactDetail',
      expect.objectContaining({
        id: expect.any(String),
      }),
    );
  });
});
