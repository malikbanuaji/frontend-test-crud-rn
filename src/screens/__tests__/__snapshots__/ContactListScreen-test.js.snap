// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList render correctly 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "height": 72,
        "padding": 16,
      }
    }
    testID="header-bar"
  >
    <Text
      style={
        Object {
          "flex": 1,
          "fontFamily": "Rubik-Light",
          "fontSize": 24,
          "marginLeft": 8,
        }
      }
    >
      KontakMu
    </Text>
  </View>
  <RCTScrollView
    ListFooterComponent={
      <View
        style={
          Object {
            "marginBottom": 28,
          }
        }
      />
    }
    data={
      Array [
        Object {
          "age": 24,
          "firstName": "Frodo",
          "id": "93ad6070-c92b-11e8-b02f-cbfa15db428b",
          "lastName": "Baggins",
          "photo": "N/A",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="contact-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "rippleRadius": undefined,
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              }
            }
            testID="contact-list-item-button"
          >
            <View
              style={
                Array [
                  Object {
                    "marginRight": 8,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#93b5e1",
                      "borderRadius": 8,
                      "height": 40,
                      "justifyContent": "center",
                      "width": 40,
                    },
                  ]
                }
              >
                <Text
                  style={
                    Object {
                      "fontFamily": "Rubik-Medium",
                      "fontSize": 14,
                    }
                  }
                >
                  FB
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column",
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontFamily": "Rubik-Medium",
                    "fontSize": 16,
                  }
                }
              >
                Frodo Baggins
              </Text>
              <Text
                style={
                  Object {
                    "fontFamily": "Rubik-Light",
                    "fontSize": 14,
                  }
                }
              >
                24
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "marginBottom": 28,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "bottom": 16,
        "position": "absolute",
        "right": 16,
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 8,
          "overflow": "hidden",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#93b5e1",
            "borderRadius": 8,
            "display": "flex",
            "flexDirection": "row",
            "overflow": "hidden",
            "padding": 12,
          }
        }
        testID="action-add-contact-button"
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 24,
              },
              Object {
                "marginRight": 4,
              },
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï†´
        </Text>
        <Text
          style={
            Object {
              "fontFamily": "Rubik-Medium",
              "fontSize": 14,
            }
          }
        >
          Tambah kontak
        </Text>
      </View>
    </View>
  </View>
</View>
`;
